tsn/src/bin/throughput.rs

```diff
@@ -0,0 +1,695 @@
+            }
+            PerfOpcode::PerfReqResult => {
+                eprintln!("Received result '{:08x}'", id);
+                tend = clock_gettime(ClockId::CLOCK_MONOTONIC).unwrap();
+                let pkt_result: PktPerfResult;
+                tdiff = tend - tstart;
+
+                pkt_info.op = PerfOpcode::PerfResResult as u8;
+                unsafe {
+                    pkt_result = PktPerfResult {
+                        pkt_count: STATS.pkt_count.to_be(),
+                        pkt_size: STATS.total_bytes.to_be(),
+                        elapsed_sec: tdiff.tv_sec(),
+                        elapsed_nsec: tdiff.tv_nsec(),
+                    };
+                }
+                let mut send_pkt = bincode::serialize(&ethernet).unwrap();
+                let mut pkt_info_bytes = bincode::serialize(&pkt_info).unwrap();
+                let mut pkt_result_bytes = bincode::serialize(&pkt_result).unwrap();
+                send_pkt.append(&mut pkt_info_bytes);
+                send_pkt.append(&mut pkt_result_bytes);
+                send_perf(sock, &mut send_pkt, size as usize);
+            }
+            _ => {
+                println!("opcode = {:0x}", opcode as u8);
+            }
+        }
+    }
+}
+
+fn statistics_thread(stat: &Statistics) {
+    let mut tdiff = TimeSpec::zero();
+    let mut start = clock_gettime(ClockId::CLOCK_MONOTONIC).unwrap();
+    let mut tlast = start;
+
+    let mut last_id: u32 = 0;
+    let mut last_pkt_count: u64 = 0;
+    let mut last_total_bytes: u64 = 0;
+
+    //TODO:let format_str = "Stat {} {} pps {} bps loss {:.3}%";
+
+    while stat.running {
+        // println!("---------Check statistic data---------");
+        let mut tnow = clock_gettime(ClockId::CLOCK_MONOTONIC).unwrap();
+        // println!("tlast = {}.{}", tlast.tv_sec(), tlast.tv_nsec());
+        // println!("tnow = {}.{}", tnow.tv_sec(), tnow.tv_nsec());
+        tsn::tsn_timespecff_diff(&mut tlast, &mut tnow, &mut tdiff);
+        // println!("tdiff after calc = {}.{}", tdiff.tv_sec(), tdiff.tv_nsec());
+
+        if tdiff.tv_sec() >= 1 {
+            tlast = tnow.clone();
+            tsn::tsn_timespecff_diff(&mut start, &mut tnow, &mut tdiff);
+            let time_elapsed: u16 = tdiff.tv_sec() as u16;
+
+            let current_pkt_count: u64 = stat.pkt_count;
+            let current_total_bytes: u64 = stat.total_bytes;
+            let current_id: u32 = stat.last_id;
+
+            // println!("current_pkt_count = {}", current_pkt_count);
+            // println!("last_pkt_count = {}", last_pkt_count);
+
+            //TODO: this code might need to be fixed
+            if last_pkt_count > current_pkt_count {
+                continue;
+            }
+
+            if last_total_bytes > current_total_bytes {
+                continue;
+            }
+
+            if last_id > current_id {
+                continue;
+            }
+            //
```

Comment:
What is this?

Reply:
The problem is if the program doesn't process `PERF_REQ_END` properly, this thread doesn't stop and keeps going. As a result, when a new client sends packets, the `last_pkt_count` variable is still storing last client's `last_pkt_count`. Therefore, when the thread is trying to calculate the data, it throws an error because `diff_pkt_count` can't hold minus value. 
The code is not sufficient and has bugs.  However, I couldn't figure out how to go about this matter. I'd love to have your input.